Source/WebCore/loader/FrameLoader.cpp

```diff
@@ -2785,16 +2787,41 @@ void FrameLoader::checkLoadCompleteForThisFrame()
             RefPtr document = m_frame->document();
             OptionSet<DataDetectorType> types = m_frame->settings().dataDetectorTypes();
             if (document && types) {
-                auto documentLevelResults = retainPtr(DataDetection::detectContentInRange(makeRangeSelectingNodeContents(*document), types, m_client->dataDetectionReferenceDate()));
-                m_frame->dataDetectionResults().setDocumentLevelResults(documentLevelResults.get());
-                if (m_frame->isMainFrame())
-                    m_client->dispatchDidFinishDataDetection(documentLevelResults.get());
+                isLoadingAsync = true;
+                DataDetection::detectContentInFrame(m_frame, types, m_client->dataDetectionReferenceDate(), [this, weakFrame = WeakPtr { m_frame.get() }, document](NSArray *results) {
+                    if (!weakFrame || &weakFrame->loader() != this)
+                        return;
+
+                    auto documentLevelResults = retainPtr(results);
+                    m_frame->dataDetectionResults().setDocumentLevelResults(documentLevelResults.get());
+                    if (m_frame->isMainFrame())
+                        m_client->dispatchDidFinishDataDetection(documentLevelResults.get());
+
+                    m_client->dispatchDidFinishLoad();
+                    auto loadingEvent = AXObjectCache::AXLoadingFinished;
+
+                    if (RefPtr document = m_frame->document()) {
+                        if (CheckedPtr cache = document->existingAXObjectCache())
+                            cache->frameLoadingEventNotification(protectedFrame().ptr(), loadingEvent);
+                    }
+
+                    if (m_frame->isMainFrame() && m_frame->page()) {
+                        ASSERT(&m_frame->page()->mainFrame() == m_frame.ptr());
+                        protectedFrame()->protectedPage()->diagnosticLoggingClient().logDiagnosticMessageWithResult(DiagnosticLoggingKeys::pageLoadedKey(), emptyString(), DiagnosticLoggingResultPass, ShouldSample::Yes);
+                    }
```

Comment:
What does this frame loading code have to do with DataDetection?

Reply:
Because originally, this code was being executed post-detection; so I wanted to maintain that order

Reply:
It would be good to understand if that is really necessary. If not, we wouldn't be taking full advantage of the async nature of the code, and potentially slowing down other things with the higher QOS.

Reply:
it works without, so doesn’t seem needed